[ADD] setup: pep518 support

sbidoul opened the pull request:
This is a non-intrusive solution for https://github.com/odoo/odoo/issues/16700, that leverages the PEP 517 build backend interface, so `pip install` and `pip wheel` work out of the box for Odoo.

sbidoul commented:
@d-fence

sbidoul commented:
@d-fence have had a chance to think about this one?

d-fence commented:
I think that the idea is really attractive but it seems that [setuptools  41](https://github.com/odoo/odoo/pull/44001/files#diff-522adf759addbd3b193c74ca85243f7dR2) is needed.
We try to stick to the Debian stable release (stretch for Odoo 12.0)  which [provides only 33](https://packages.debian.org/search?keywords=python3%2Dsetuptools).
The same idea applies to the [included packaging script](https://github.com/odoo/odoo/blob/12.0/setup/package.py) that uses a [Dockerfile](https://github.com/odoo/odoo/blob/12.0/setup/package.dfdebian) corresponding to the current Debian version at the time of the release. I'm afraid that this PR could break the actual packaging system.
We could merge that into master when bullseye become stable.

sbidoul commented:
@d-fence thanks for looking into this.

Actually the setuptools 41 requirement is the one that has the `build_meta` PEP 517 API, but it is not needed to run Odoo.

When running `pip install .` or `pip wheel .` with a pip version that supports PEP 517, pip will download the required build requirements from PyPI in an isolated environment to run the build.

So this should be safe to merge, independently of the distro version for pip and setuptools.

sbidoul commented:
@d-fence would you consider this for v14?

d-fence commented:
Debian buster only [provides ](https://packages.debian.org/buster/python3-setuptools) setup-tools 40.
41 is needed as I read.

sbidoul commented:
But as I mentioned above, this is not needed at runtime nor for debian packaging. And even on buster `pip install` will download the required setuptools version from PyPI.

sbidoul commented:
@d-fence do you thing this can be considered for v15 ?
What are your remaining concerns ?
If you agree to go ahead I'll do a new pass on this and polish the PR.

simahawk commented:
@d-fence any update?

sbidoul commented:
I cleaned-up and modernized the code, and made it more robust.

sbidoul commented:
@d-fence any chance this can be considered for v16 ? I retargeted this branch to master.

rousseldenis commented:
@d-fence gentle reminder. Thanks

yajo commented:
I'm getting plenty of warnings like this when building Odoo with this PR enabled:

```
python3.8-odoo>  SetuptoolsDeprecationWarning:     Installing 'odoo.tools.data.files' as data is deprecated, please list it in `packages`.
python3.8-odoo>       !!
python3.8-odoo>       ############################
python3.8-odoo>       # Package would be ignored #
python3.8-odoo>       ############################
python3.8-odoo>       Python recognizes 'odoo.tools.data.files' as an importable package,
python3.8-odoo>       but it is not listed in the `packages` configuration of setuptools.
python3.8-odoo>       'odoo.tools.data.files' has been automatically added to the distribution only
python3.8-odoo>       because it may contain data files, but this behavior is likely to change
python3.8-odoo>       in future versions of setuptools (and therefore is considered deprecated).
python3.8-odoo>       Please make sure that 'odoo.tools.data.files' is included as a package by using
python3.8-odoo>       the `packages` configuration field or the proper discovery methods
python3.8-odoo>       (for example by using `find_namespace_packages(...)`/`find_namespace:`
python3.8-odoo>       instead of `find_packages(...)`/`find:`).
python3.8-odoo>       You can read more about "package discovery" and "data files" on setuptools
python3.8-odoo>       documentation page.
python3.8-odoo>   !!
python3.8-odoo>     check.warn(importable)
python3.8-odoo>   Building wheel for odoo (pyproject.toml) ... done
```

Should you add some method to find packages that is more future-proof? 🤔

KangOl commented on setup/pep517_odoo.py:
```diff
@@ -0,0 +1,62 @@
+                    addon_path.symlink_to(
+                        link_target / target_addon_path.name, target_is_directory=True
+                    )
+                symlinks.append(addon_path)
```
```suggestion
                    symlinks.append(addon_path)
```

KangOl reviewed (changes requested):
We should not unlink existing symlink

sbidoul commented:
@KangOl thanks for your review! I updated and rebased on 17 as requested.
pylint complains on the `import *` but I think it's the right way to do in this case, as not all setuptools version implement the same build hooks.

yajo commented:
Here's [the patch for Odoo 16](https://github.com/odoo/odoo/files/14083526/16.0.75b0730.patch.txt), for those that still need it there.

sbidoul commented:
@KangOl @d-fence gentle nudge here.

sbidoul commented:
@KangOl @d-fence this PR is becoming more pressing. Indeed there are more and more situations where pip defaults to doing a PEP 660 editable install which triggers a setuptools installation mode that is not compatible with Odoo. It is therefore important to force the `compat` editable mode, as this PR does.

Pretty please :pray:

simahawk commented:
@KangOl @d-fence any update? :)

gustavovalverde commented:
Seems like the `master` version will break the `pip install -e` approach

In Odoo 18.0:

```bash
2025-04-16 08:47:17,780 1 INFO ? odoo: Odoo version 18.0 
2025-04-16 08:47:17,780 1 INFO ? odoo: Using configuration file at /etc/odoo/odoo.conf 
2025-04-16 08:47:17,780 1 INFO ? odoo: addons paths: ['/opt/odoo/odoo/addons', '/var/lib/odoo/data/addons/18.0', '/opt/odoo/addons']
```

In master (as of 04/16/2025)

```bash
2025-04-16 08:47:25,243 1 INFO ? odoo: Odoo version 18.3a1 
2025-04-16 08:47:25,243 1 INFO ? odoo: Using configuration file at /etc/odoo/odoo.conf 
2025-04-16 08:47:25,243 1 INFO ? odoo: addons paths: _NamespacePath(['/opt/odoo/odoo/addons', '/opt/odoo/odoo/addons', '__editable__.odoo-18.3a1.finder.__path_hook__', '/var/lib/odoo/data/addons/18.3', '/opt/odoo/addons']) 
2025-04-16 08:47:26,126 1 WARNING test odoo.modules.module: addons path does not exist: __editable__.odoo-18.3a1.finder.__path_hook__ 
```

This will be critical in the next release
